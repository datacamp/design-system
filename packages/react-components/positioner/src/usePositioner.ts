import { css, SerializedStyles } from '@emotion/react';
import React from 'react';

const usePositioner = ({
  isVisible,
  offset = 0,
  position,
  target,
}: {
  isVisible: boolean;
  offset?: number;
  position:
    | 'bottom'
    | 'top'
    | 'right'
    | 'left'
    | 'bottomLeft'
    | 'bottomRight'
    | 'topLeft'
    | 'topRight';
  target: React.MutableRefObject<HTMLElement | undefined | null>;
}): SerializedStyles => {
  const [targetBox, setTargetBox] = React.useState<{
    bottom: number;
    height: number;
    left: number;
    right: number;
    top: number;
    width: number;
  } | null>(null);

  React.useEffect(() => {
    const resetPosition = (): void => {
      if (!isVisible || !target.current) return;

      setTargetBox(target.current.getBoundingClientRect());
    };

    resetPosition();
    window.addEventListener('scroll', resetPosition, true);
    window.addEventListener('resize', resetPosition, true);

    return () => {
      window.removeEventListener('scroll', resetPosition, true);
      window.removeEventListener('resize', resetPosition, true);
    };
  }, [isVisible, target]);

  if (!targetBox || !isVisible) return css({ display: 'none' });

  const layoutCSSMap = {
    bottom: css({
      left: (targetBox.left + targetBox.right) / 2,
      top: targetBox.bottom + offset,
      transform: 'translate(-50%, 0)',
    }),
    bottomLeft: css({
      left: targetBox.left,
      top: targetBox.bottom + offset,
    }),
    bottomRight: css({
      left: targetBox.left + targetBox.width,
      top: targetBox.bottom + offset,
      transform: 'translate(-100%)',
    }),
    left: css({
      left: targetBox.left - offset,
      top: targetBox.top + targetBox.height / 2,
      transform: 'translate(-100%, -50%)',
    }),
    right: css({
      left: targetBox.right + offset,
      top: targetBox.top + targetBox.height / 2,
      transform: 'translate(0, -50%)',
    }),
    top: css({
      left: (targetBox.left + targetBox.right) / 2,
      top: targetBox.top - offset,
      transform: 'translate(-50%, -100%)',
    }),
    topLeft: css({
      left: targetBox.left,
      top: targetBox.top - offset,
      transform: 'translate(0, -100%)',
    }),
    topRight: css({
      left: targetBox.left + targetBox.width,
      top: targetBox.top - offset,
      transform: 'translate(-100%, -100%)',
    }),
  };

  return css(
    {
      position: 'fixed',
    },
    layoutCSSMap[position],
  );
};

export default usePositioner;
