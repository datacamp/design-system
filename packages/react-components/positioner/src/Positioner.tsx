import { css } from '@emotion/react';
import React, { MutableRefObject } from 'react';

import usePositioner from './usePositioner';

interface PositionerProps {
  /**
   * The content
   */
  children: React.ReactNode;
  position:
    | 'bottom'
    | 'top'
    | 'right'
    | 'left'
    | 'bottomLeft'
    | 'bottomRight'
    | 'topLeft'
    | 'topRight';
  target: MutableRefObject<HTMLElement | undefined | null>;
  visible: boolean;
}

const Positioner = ({
  children,
  position,
  target,
  visible,
}: PositionerProps): React.ReactElement | null => {
  const style = usePositioner({ isVisible: visible, position, target });

  return visible ? <div css={css(style, { zIndex: 1 })}>{children}</div> : null;
};

export default Positioner;
