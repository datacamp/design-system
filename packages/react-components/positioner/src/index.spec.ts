import Positioner from './Positioner';
import usePositioner from './usePositioner';

import ExportedPositioner, { usePositioner as exportedUsePositioner } from '.';

describe('exports', () => {
  it('exports the right thing', () => {
    expect(ExportedPositioner).toEqual(Positioner);
    expect(exportedUsePositioner).toEqual(usePositioner);
  });
});
