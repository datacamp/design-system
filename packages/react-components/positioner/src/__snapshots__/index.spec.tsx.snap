// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Positioner renders on the bottom on the target 1`] = `
.emotion-0 {
  height: 50px;
  width: 50px;
}

.emotion-1 {
  height: 10px;
  width: 10px;
}

.emotion-2 {
  position: absolute;
  left: 0;
  top: 0;
  -webkit-transform: translate(-50%,0);
  -ms-transform: translate(-50%,0);
  transform: translate(-50%,0);
}

<body>
  <div>
    <div
      class="emotion-0"
      data-test-id="target"
    />
  </div>
  <div
    class="emotion-2"
  >
    <div
      class="emotion-1"
      data-test-id="element"
    />
  </div>
</body>
`;

exports[`Positioner renders on the bottomLeft on the target 1`] = `
.emotion-0 {
  height: 50px;
  width: 50px;
}

.emotion-2 {
  position: absolute;
  left: 0;
  top: 0;
}

.emotion-1 {
  height: 10px;
  width: 10px;
}

<body>
  <div>
    <div
      class="emotion-0"
      data-test-id="target"
    />
  </div>
  <div
    class="emotion-2"
  >
    <div
      class="emotion-1"
      data-test-id="element"
    />
  </div>
</body>
`;

exports[`Positioner renders on the bottomRight on the target 1`] = `
.emotion-0 {
  height: 50px;
  width: 50px;
}

.emotion-1 {
  height: 10px;
  width: 10px;
}

.emotion-2 {
  position: absolute;
  left: 0;
  top: 0;
  -webkit-transform: translate(-100%);
  -ms-transform: translate(-100%);
  transform: translate(-100%);
}

<body>
  <div>
    <div
      class="emotion-0"
      data-test-id="target"
    />
  </div>
  <div
    class="emotion-2"
  >
    <div
      class="emotion-1"
      data-test-id="element"
    />
  </div>
</body>
`;

exports[`Positioner renders on the left on the target 1`] = `
.emotion-0 {
  height: 50px;
  width: 50px;
}

.emotion-1 {
  height: 10px;
  width: 10px;
}

.emotion-2 {
  position: absolute;
  left: 0;
  top: 0;
  -webkit-transform: translate(-100%,-50%);
  -ms-transform: translate(-100%,-50%);
  transform: translate(-100%,-50%);
}

<body>
  <div>
    <div
      class="emotion-0"
      data-test-id="target"
    />
  </div>
  <div
    class="emotion-2"
  >
    <div
      class="emotion-1"
      data-test-id="element"
    />
  </div>
</body>
`;

exports[`Positioner renders on the right on the target 1`] = `
.emotion-0 {
  height: 50px;
  width: 50px;
}

.emotion-1 {
  height: 10px;
  width: 10px;
}

.emotion-2 {
  position: absolute;
  left: 0;
  top: 0;
  -webkit-transform: translate(0,-50%);
  -ms-transform: translate(0,-50%);
  transform: translate(0,-50%);
}

<body>
  <div>
    <div
      class="emotion-0"
      data-test-id="target"
    />
  </div>
  <div
    class="emotion-2"
  >
    <div
      class="emotion-1"
      data-test-id="element"
    />
  </div>
</body>
`;

exports[`Positioner renders on the top on the target 1`] = `
.emotion-0 {
  height: 50px;
  width: 50px;
}

.emotion-1 {
  height: 10px;
  width: 10px;
}

.emotion-2 {
  position: absolute;
  left: 0;
  top: 0;
  -webkit-transform: translate(-50%,-100%);
  -ms-transform: translate(-50%,-100%);
  transform: translate(-50%,-100%);
}

<body>
  <div>
    <div
      class="emotion-0"
      data-test-id="target"
    />
  </div>
  <div
    class="emotion-2"
  >
    <div
      class="emotion-1"
      data-test-id="element"
    />
  </div>
</body>
`;

exports[`Positioner renders on the topLeft on the target 1`] = `
.emotion-0 {
  height: 50px;
  width: 50px;
}

.emotion-1 {
  height: 10px;
  width: 10px;
}

.emotion-2 {
  position: absolute;
  left: 0;
  top: 0;
  -webkit-transform: translate(0,-100%);
  -ms-transform: translate(0,-100%);
  transform: translate(0,-100%);
}

<body>
  <div>
    <div
      class="emotion-0"
      data-test-id="target"
    />
  </div>
  <div
    class="emotion-2"
  >
    <div
      class="emotion-1"
      data-test-id="element"
    />
  </div>
</body>
`;

exports[`Positioner renders on the topRight on the target 1`] = `
.emotion-0 {
  height: 50px;
  width: 50px;
}

.emotion-1 {
  height: 10px;
  width: 10px;
}

.emotion-2 {
  position: absolute;
  left: 0;
  top: 0;
  -webkit-transform: translate(-100%,-100%);
  -ms-transform: translate(-100%,-100%);
  transform: translate(-100%,-100%);
}

<body>
  <div>
    <div
      class="emotion-0"
      data-test-id="target"
    />
  </div>
  <div
    class="emotion-2"
  >
    <div
      class="emotion-1"
      data-test-id="element"
    />
  </div>
</body>
`;
